src/dawn/Optimizer/PassFieldVersioning.cpp

```diff
@@ -89,71 +91,165 @@ static void reportRaceCondition(const Statement& statement,
+            auto rc = fixRaceCondition(stencilInstantiation, newGraph.get(), stencil, doMethod,
+                                       loopOrder, stageIdx, stmtIndex);
+
+            if(rc == RCKind::RK_Unresolvable) {
+              // Nothing we can do ... bail out
+              return false;
+            } else if(rc == RCKind::RK_Fixed) {
+              // We fixed a race condition (this means some fields have changed and our current
+              // graph
+              // is invalid)
+              newGraph = oldGraph;
+              newGraph->insertStatementAccessesPair(stmtAccessesPair);
+            }
+            doMethod.update(iir::NodeUpdateType::level);
           }
-          doMethod.update(iir::NodeUpdateType::level);
+          stage.update(iir::NodeUpdateType::level);
         }
-        stage.update(iir::NodeUpdateType::level);
+        stageIdx--;
+      }
+
+      for(const auto& ms : iterateIIROver<iir::MultiStage>(stencil)) {
+        ms->update(iir::NodeUpdateType::levelAndTreeAbove);
       }
-      stageIdx--;
     }
 
-    for(const auto& ms : iterateIIROver<iir::MultiStage>(stencil)) {
-      ms->update(iir::NodeUpdateType::levelAndTreeAbove);
+    if(context->getOptions().ReportPassFieldVersioning && numRenames_ == 0)
+      std::cout << "\nPASS: " << getName() << ": " << stencilInstantiation->getName()
+                << ": no rename\n";
+    return true;
+  } else if(mode_ == FieldVersioningPassMode::FM_FixAccess) {
+    for(int id : stencilInstantiation->getMetaData()
+                     .getFieldAccessMetadata()
+                     .getVariableVersions()
+                     .getVersionIDs()) {
+      for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
+        iir::Stencil& stencil = *stencilPtr;
+        for(const auto& mss : iterateIIROver<iir::MultiStage>(stencil)) {
+          // For every mulitstage, check if the versioned field as a read-access first
+          auto multiInterval = mss->computeReadAccessInterval(id);
+          if(multiInterval.empty())
+            continue;
+
+          // If it has one, we need to generate a domethod that fills that access from the
+          // original field in every interval that is being accessed before it's first write
+          auto extents = mss->getField(id).getReadExtentsRB();
+          for(auto interval : multiInterval.getIntervals()) {
+
+            // we create the new stage that holds these do-methods
+            auto insertedStage = createAssignmentStage(
+                interval, id, stencilInstantiation->getMetaData().getOriginalVersionOfAccessID(id),
+                stencilInstantiation, extents);
+            // and insert them at the beginnning of the MultiStage
+            mss->insertChild(mss->childrenBegin(), std::move(insertedStage));
+          }
+          // update the mss: #TODO: this is still a workaround since we don't have level-and below:
```

Comment:
I dont understand the comment, in principle the derived info of node levels below does not depend on parent. So we dont need an update of level and below, isnt?

Reply:
In this specific case we modify the children of an instance. Now the ideal thing would be:
- Update all the Do-Methods
- Update all the above (once)

Now we have level-and-tree above for Do-Methods but this recomputes the the derived info multiple times, doesn't it?

So with the current tools the perfect way would be:
iteateIIROver<DoMethod>
  update(level)
iteateIIROver<Stage>
  update(level)
MS->update(levelAndTreeAbove)

now the question is if we want to wrap this into a function (like update-range (From DoMethod to MS) or have one that does MS and below (which does level on every child so we do it only once).

Thoughts?

Reply:
yeah, ok I see, 
the API was not designed for updating efficiently when all the leaves are updated, 
ok, I would leave it like that then